src/Http/Routing/src/Matching/NegotiationMatcherPolicy.cs

```diff
@@ -0,0 +1,443 @@
+    // * Note that the priority of the route applies first, that is, in the last case, for a request to Resource.css with no Accept-Encoding header, the Resource.css (2)
+    //   endpoint will still be selected over the catchall endpoint.
+    public IReadOnlyList<PolicyNodeEdge> GetEdges(IReadOnlyList<Endpoint> endpoints)
+    {
+        ArgumentNullException.ThrowIfNull(endpoints);
+
+        // The algorithm here is designed to be preserve the order of the endpoints
+        // while also being relatively simple. Preserving order is important.
+
+        // First, build a dictionary of all of the content-type patterns that are included
+        // at this node.
+        //
+        // For now we're just building up the set of keys. We don't add any endpoints
+        // to lists now because we don't want ordering problems.
+        var edges = new Dictionary<string, List<Endpoint>>(StringComparer.OrdinalIgnoreCase);
+        for (var i = 0; i < endpoints.Count; i++)
+        {
+            var endpoint = endpoints[i];
+            var metadata = GetMetadataValue(endpoint) ?? DefaultNegotiationValue;
+            if (!edges.TryGetValue(metadata, out var endpointsForType))
+            {
+                edges.Add(metadata, []);
+            }
+        }
+
+        // Now in a second loop, add endpoints to these lists.
+        // We've enumerated all of the states, so we want to see which states each endpoint matches.
+        for (var i = 0; i < endpoints.Count; i++)
+        {
+            var endpoint = endpoints[i];
+            var metadata = GetMetadataValue(endpoint) ?? DefaultNegotiationValue;
+            if (string.Equals(metadata, DefaultNegotiationValue, StringComparison.OrdinalIgnoreCase))
+            {
+                // This means that this endpoint does not specify a negotiation value, a default of identity is assumed.
+                // Which means that this endpoint is always a candidate.
+                foreach (var edge in edges)
+                {
+                    edge.Value.Add(endpoint);
+                }
+            }
+            else
+            {
+                var endpointsForType = edges[metadata];
+                endpointsForType.Add(endpoint);
+            }
+        }
+
+        // If after we're done there isn't any endpoint that accepts the default encoding, then we'll synthesize an
+        // endpoint that always returns a 406.
+        if (!edges.TryGetValue(DefaultNegotiationValue, out var anyEndpoints))
+        {
+            anyEndpoints = [CreateRejectionEndpoint()];
+            edges.Add(DefaultNegotiationValue, anyEndpoints);
+
+            // Add a node to use when there is no negotiation header.
+            edges.Add(string.Empty, anyEndpoints);
+        }
+        else
+        {
+            // If there is an endpoint that accepts an then it is also used when there is no content type
+            edges.Add(string.Empty, anyEndpoints);
+        }
```

Comment:
Really appreciate all the detailed comments in this PR explaining background context and documenting the reasoning behind the intricate parts of the implementation!